moo.js

```diff
@@ -36,6 +36,22 @@
     return b.length - a.length
   }
 
+  function sortPatterns(array) {
+    var regexps = []
+    var literals = []
+    // sort literals by length to ensure longest match
+    for (var i=0; i<array.length; i++) {
+      var obj = array[i]
+      ;(isRegExp(obj) ? regexps : literals).push(obj)
+    }
+    literals.sort(compareLength)
+    // append regexps to the end
+    for (var i=0; i<regexps.length; i++) {
+      literals.push(regexps[i])
+    }
+    return literals
```

Comment:
Maybe just `literals.concat(regexps)` (or `literals.push.apply(literals, regexps)` and get rid of the loop?